ecs/pod_creator.go

```diff
@@ -27,11 +31,227 @@ func NewBasicECSPodCreator(c cocoa.ECSClient, v cocoa.Vault) (*BasicECSPodCreato
+		for _, failure := range runOut.Failures {
+			err = errors.Errorf("task '%s': %s: %s\n", *failure.Arn, *failure.Detail, *failure.Reason)
+			catcher.Add(err)
+		}
+		return nil, errors.Wrap(catcher.Resolve(), "running task")
+	}
+
+	if len(runOut.Tasks) == 0 || runOut.Tasks[0].TaskArn == nil {
+		return nil, errors.New("missing running task")
+	}
+
+	resources := cocoa.NewECSPodResources().
+		SetCluster(*mergedPodExecutionOpts.Cluster).
+		SetSecrets(allSecrets).
+		SetTaskDefinition(*taskDef).
+		SetTaskID(*runOut.Tasks[0].TaskArn)
+
+	options := NewBasicECSPodOptions().
+		SetClient(m.client).
+		SetVault(m.vault).
+		SetStatus(cocoa.Running).
+		SetResources(*resources)
+
+	p, err := NewBasicECSPod(options)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "creating pod")
+	}
+
+	return p, nil
 }
 
 // CreatePodFromExistingDefinition creates a new pod backed by AWS ECS from an
 // existing definition.
 func (m *BasicECSPodCreator) CreatePodFromExistingDefinition(ctx context.Context, def cocoa.ECSTaskDefinition, opts ...*cocoa.ECSPodExecutionOptions) (cocoa.ECSPod, error) {
 	return nil, errors.New("TODO: implement")
 }
+
+// translateStringArrayToECSTagArray translates the strings into ECS tags.
+func translateStringArrayToECSTagArray(tags []string) []*ecs.Tag {
+	ecsTags := []*ecs.Tag{}
+	for _, tag := range tags {
+		ecsTag := &ecs.Tag{}
+		ecsTag.SetKey(tag)
+		ecsTags = append(ecsTags, ecsTag)
+	}
+	return ecsTags
+}
+
+// translateStrategy translates the strategy and parameter into ECS placement strategy.
+func translateStrategy(strategy *cocoa.ECSPlacementStrategy, param *cocoa.ECSStrategyParameter) []*ecs.PlacementStrategy {
+	placementStrat := ecs.PlacementStrategy{}
+	placementStrat.SetType(string(*strategy)).SetField(*param)
+	return []*ecs.PlacementStrategy{&placementStrat}
+}
+
+// exportEnvVars translates the environment variables into ECS environment variables and secrets.
+func (m *BasicECSPodCreator) exportEnvVars(ctx context.Context, variables []cocoa.EnvironmentVariable) ([]*ecs.KeyValuePair, []cocoa.PodSecret, []cocoa.NamedSecret, error) {
+	keyValuePairs := []*ecs.KeyValuePair{}
+	secrets := []cocoa.PodSecret{}
+	createSecrets := []cocoa.NamedSecret{}
+
+	for _, variable := range variables {
+		if variable.SecretOpts == nil {
+			keyValue := ecs.KeyValuePair{}
+			keyValue.SetName(*variable.Name).SetValue(*variable.Value)
+			keyValuePairs = append(keyValuePairs, &keyValue)
+		} else {
+			if !*variable.SecretOpts.Exists {
+				createSecrets = append(createSecrets, variable.SecretOpts.PodSecret.NamedSecret)
+			}
+			secrets = append(secrets, variable.SecretOpts.PodSecret)
+		}
+	}
+
+	return keyValuePairs, secrets, createSecrets, nil
+}
+
+// exportSecrets extracts all secrets and environment variables from ECS creation options
+func (m *BasicECSPodCreator) exportSecrets(ctx context.Context, merged cocoa.ECSPodCreationOptions) ([]*ecs.KeyValuePair, []cocoa.PodSecret, []cocoa.NamedSecret, error) {
```

Comment:
This isn't really exporting anything except for the environment variables. As a general rule, "exporting" something usually means converting it to a format which is suitable for an external API (e.g. ECS), but this is mostly creating internal data structures (i.e. cocoa secret structs).

Reply:
I would split these 3 return values into different functions so that `exportSecrets` doesn't extract all secrets and environment variables - it should only export the secrets. Naming a function one way and doing something different in the actual code (i.e. extracting both secrets and non-secrets) is misleading and causes confusion when someone else has to read it.

See my comment about `exportEnvVars`.